### PR TITLE
Tech/react wysiwsyg

### PR DESCRIPTION
**Pull request by florianduros**
Migrating the vanilla js web to react

There is still TODO part in the code but I think the current state of this PR is reviewable (and maybe soon mergeable).

- 3 tests on special characters don't work , I think is due to of the jsdom environnement (I'm not sure)
- The code coverage is near 55%, I just migrated the old web tests in it, it will need a dedicated PR to strengthen in
- The autofocus feature can be improved
- In an other PR, the packaging for npm and the live demo will be refined.

### CODE REVIEW

**Review by andybalaam — APPROVED** [review]
Looks awesome. A few small nitpicks, but feel free to merge when ready!

File: platforms/web/src/README.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    - yarn
+
+
+```sh
+yarn install
+```
+
+### Build
+
+[Vite](https://vitejs.dev/) is the Wysiwyg Composer builder.
```
[the remark this answers is not in the document]
Comment (florianduros) [reply platforms/web/src/README.md:30]: Yes, it's not a prerequisite.

Vite is downloaded during the `yarn install` command.

It's just extra information

File: platforms/web/src/README.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+#### Generate WASM bindings
+
+The composer uses a cross-platform rust library. In order to work as intended, the WASM bindings must be generate according to the [Matrix Wysisyg README.md](../../../README.md)
+
+#### Yarn install
+
+Requirements:
+    - node
+    - yarn
```
[the remark this answers is not in the document]
Comment (florianduros) [reply platforms/web/src/README.md:21]: Yes, I will added:

For node, I used node 18, I don't if it works in lesser version.
For yarn, I used the same than our sdk and element web, so I kept 1.x

File: platforms/web/src/src/lib/dom.test.ts [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+        // When
+        offset = computeSelectionOffset(editor.childNodes[0], 20);
+
+        // Then
+        expect(offset).toBe(20);
+    });
+});
+
+describe('countCodeunit', () => {
+    test('Should count ASCII', () => {
```
[the remark this answers is not in the document]
Comment (florianduros) [reply platforms/web/src/src/lib/dom.test.ts:207]: Good point

File: platforms/web/src/src/lib/useFormattingActions.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        const sendFormatBlockEvent = (e: ReactMouseEvent<HTMLElement, MouseEvent>, blockType: BlockType) => {
+            e.preventDefault();
+            e.stopPropagation();
+            editorRef.current?.dispatchEvent(new CustomEvent('formatBlock', { detail: { blockType } }));
+        };
+
+        /*return Object.keys(actions).reduce<FormattingActions>((acc, action) => {
+            acc[action] = (e: MouseEvent) => sendInputEvent(e, actions[action]);
+            return acc;
+        }, {} as FormattingActions);*/
```
[the remark this answers is not in the document]
Comment (florianduros) [reply platforms/web/src/src/lib/useFormattingActions.ts:33]: I have an other way to generate the actions.

The commented one is more 'automatic' but a part of the TS completion is lost. 

The other one is more declarative, more copy/paste but TS works fine in it.

File: Makefile [hunk 5]
```diff
@@ -83,4 +90,4 @@ test:
 coverage:
 	@echo "Requires `rustup component add llvm-tools-preview`"
 	@echo "Requires `cargo install cargo-llvm-cov`"
-	cargo llvm-cov --open
+	cargo llvm-cov --open
```
[the remark this answers is not in the document]
Comment (florianduros) [reply Makefile]: Done :)

**Pull request merged**